### PR TITLE
Add parametrization version of weight_norm

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103009
* __->__ #103001

This done in the ordinary way, but also:

* Deprecation warning for the old API, and a migration guide
* Backwards compatibility for state_dict loading the old weight_norm
* Test for pickling and deepcopy, which was the motivating reason

weight_norm is still used by HuggingFace Wav2Vec2.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torch/nn/utils/parametrizations.py [hunk 1]
```diff
@@ -285,6 +285,84 @@ def orthogonal(module: Module,
     return module
 
 
+class _WeightNorm(Module):
+    def __init__(
+        self,
+        dim: int = 0,
```
Comment (Skylion007) [inline torch/nn/utils/parametrizations.py:291]:
```suggestion
        dim: Optional[int] = 0,
```

**Comment by ezyang** [comment]
cc @vadimkantorov

File: torch/nn/utils/weight_norm.py [hunk 2]
```diff
@@ -87,6 +90,27 @@ def weight_norm(module: T_module, name: str = 'weight', dim: int = 0) -> T_modul
 
     See https://arxiv.org/abs/1602.07868
 
+    .. warning::
+
+        This function is deprecated.  Use :func:`torch.nn.utils.parametrizations.weight_norm`
+        which uses the modern parametrization API.  The new ``weight_norm`` is compatible
+        with ``state_dict`` generated from old ``weight_norm``.
+
+        Migration guide:
```
Comment (lezcano) [inline torch/nn/utils/weight_norm.py:99]: 10/10 migration guide.

File: torch/nn/utils/parametrizations.py [hunk 3]
```diff
@@ -285,6 +285,84 @@ def orthogonal(module: Module,
...
+        torch.Size([40, 1])
+        >>> m.parametrizations.weight.original1.size()
+        torch.Size([40, 20])
+
+    """
+    weight = getattr(module, name, None)
+    if not isinstance(weight, Tensor):
+        raise ValueError(
+            "Module '{}' has no parameter or buffer with name '{}'".format(module, name)
+        )
```
Comment (lezcano) [inline torch/nn/utils/parametrizations.py:349]: Not needed, this is already performed in `register_parametrization`. Even more, if you want to parametrize an already parametrized weight, this would trigger its forward unnecessarily.

File: torch/nn/utils/parametrizations.py [hunk 4]
```diff
@@ -285,6 +285,84 @@ def orthogonal(module: Module,
...
+            dim = -1
+        self.dim = dim
+
+    def forward(self, weight_g, weight_v):
+        return torch._weight_norm(weight_v, weight_g, self.dim)
+
+    def right_inverse(self, weight):
+        # TODO: is the .data necessary?
+        weight_g = torch.norm_except_dim(weight, 2, self.dim).data
+        weight_v = weight.data
```
Comment (lezcano) [inline torch/nn/utils/parametrizations.py:304]: Pretty sure it's not necessary in the first one. Not sure about the second one as it's written.

That being said, `weight_v` shoudl be `weight` divided by `weight_g`. When that's written that way, then I'm pretty sure you don't need the `.data`.

Comment (ezyang) [reply torch/nn/utils/parametrizations.py:304]: Hmm, but that is not what the original code does?

```
        module.register_parameter(name + '_g', Parameter(norm_except_dim(weight, 2, dim).data))
        module.register_parameter(name + '_v', Parameter(weight.data))
        setattr(module, name, fn.compute_weight(module))
```

Comment (ezyang) [reply torch/nn/utils/parametrizations.py:304]: I reread the paper and not dividing is correct. What is confusing in the docs is we refer to a direction vector, but we do not actually represent the direction vector directly (because it is required to orthogonal) so classic weight norm also reparametrizes the direction vector into v (the original weight vector) with the direction vector now represented as `v / || v ||`

File: torch/nn/utils/parametrizations.py [hunk 5]
```diff
@@ -285,6 +285,84 @@ def orthogonal(module: Module,
...
+    .. math::
+         \mathbf{w} = g \dfrac{\mathbf{v}}{\|\mathbf{v}\|}
+
+    Weight normalization is a reparameterization that decouples the magnitude
+    of a weight tensor from its direction. This replaces the parameter specified
+    by :attr:`name` with two parameters: one specifying the magnitude
+    and one specifying the direction.
+
+    By default, with ``dim=0``, the norm is computed independently per output
+    channel/plane. To compute a norm over the entire weight tensor, use
```
Comment (albanD) [inline torch/nn/utils/parametrizations.py:320]: I know this was here before but dim=None -> dim=-1 which computes the norm over the last dimension no?

**Review by albanD — APPROVED** [review]
Thanks!

File: torch/nn/utils/parametrizations.py (hunk 5 above)
Comment (ezyang) [reply torch/nn/utils/parametrizations.py:320]: No, I believe weight_norm is special cased for this weird behavior lol.

```
>>> v = torch.randn(2, 2)
>>> g = torch.randn(())
>>> torch._weight_norm(v, g, 0)
tensor([[ 0.0577, -0.0345],
        [-0.0000,  0.0000]])
>>> torch._weight_norm(v, g, 1)
tensor([[0.0589, 0.0000],
        [0.0325, 0.0000]])
>>> torch._weight_norm(v, g, -1)
tensor([[ 0.0489, -0.0293],
        [ 0.0270, -0.0233]])
```

Comment (vadimkantorov) [reply torch/nn/utils/parametrizations.py:320]: May also be a good moment to eliminate older implicit behavior? and maybe simply require passing explicit dim?

File: torch/nn/utils/parametrizations.py [hunk 6]
```diff
@@ -285,6 +285,84 @@ def orthogonal(module: Module,
...
+        >>> m.parametrizations.weight.original0.size()
+        torch.Size([40, 1])
+        >>> m.parametrizations.weight.original1.size()
+        torch.Size([40, 20])
+
+    """
+    _weight_norm = _WeightNorm(dim)
+    parametrize.register_parametrization(module, name, _weight_norm, unsafe=True)
+
+    def _weight_norm_compat_hook(state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs):
```
Comment (vadimkantorov) [inline torch/nn/utils/parametrizations.py:354]: also, should installation of this hook be optional? (to avoid surprising interaction / influence with other hooks)

File: torch/nn/utils/parametrizations.py (hunk 4 above)
Comment (lezcano) [reply torch/nn/utils/parametrizations.py]: Right, had a look at the paper and indeed it's as you mention.

File: torch/nn/utils/parametrizations.py (hunk 6 above)
Comment (vadimkantorov) [reply torch/nn/utils/parametrizations.py:354]: And maybe still remove the underscore from _WeightNorm?

Comment (ezyang) [reply torch/nn/utils/parametrizations.py:354]: @lezcano can you explain to us why the preexisting parametrizations are underscore'd on their module?

Comment (lezcano) [reply torch/nn/utils/parametrizations.py:354]: The idea is that they are not exposed as public API (that API is not documented) hence the underscore.

Comment (vadimkantorov) [reply torch/nn/utils/parametrizations.py:354]: I guess the question is why are they not exposed... some time has passed since their introduction, so maybe now could make them public api

Comment (vadimkantorov) [reply torch/nn/utils/parametrizations.py:354]: To expose more the parameterization design to general public (so that people would take the WeightNorm code as code example - and I think it's mature enough for this), signal that it's mature, to encourage less use of "magical" methods `register_weight_norm` and expose users to compositionality of parametrizations.

If these are public, the current `register_weight_norm` and its friends can be deprecated.

Comment (ezyang) [reply torch/nn/utils/parametrizations.py:354]: I must admit, I do find making the Parametrization modules private a bit odd. We don't make normal nn Modules private; why should parametrizations be different? (That being said I admit the precautionary principle says that you should opt to make things private if there is no reason for them to be public.)

Comment (vadimkantorov) [reply torch/nn/utils/parametrizations.py:354]: Yes, but I propose that people directly use `parametrize.register_parametrization` for that. And deprecating the special method `weight_norm(...)` and asking to directly use `parametrize.register_parametrization` would highlight this compositionality - especially if this helper now does nothing except create a `_WeightNorm` and call `parametrize.register_parametrization`

Comment (vadimkantorov) [reply torch/nn/utils/parametrizations.py:354]: I think in earlier days making them private was more justified because it was the move to the new parameterization interface (and the old way was calling a helper method), but now it's more stable and it makes sense to promote the direct use of parametrization modules like `_WeightNorm`...

Comment (lezcano) [reply torch/nn/utils/parametrizations.py:354]: I do prefer these helper methods. I did something similar in my library https://github.com/Lezcano/geotorch/, and I very much prefer the functional API than the one where you use `register_parametrization` directly. But yeah, this is very subjective indeed.

Comment (vadimkantorov) [reply torch/nn/utils/parametrizations.py:354]: Btw `register_parametrization` seems to return `self`, so could be used in quasi-functional way as well! :)
